packages/go_router/lib/src/route.dart

```diff
@@ -458,4 +540,375 @@ class ShellRoute extends RouteBase {
+/// configure [StatefulShellRoute].
+///
+/// The only required argument when creating a StatefulShellBranch is the
+/// sub-routes ([routes]), however sometimes it may be convenient to also
+/// provide a [initialLocation]. The value of this parameter is used when
+/// loading the branch for the first time (for instance when switching branch
+/// using the goBranch method in [StatefulShellBranchState]). A [navigatorKey]
+/// can be useful to provide in case it's necessary to access the [Navigator]
+/// created for this branch elsewhere.
+@immutable
+class StatefulShellBranch {
+  /// Constructs a [StatefulShellBranch].
+  StatefulShellBranch({
+    required this.routes,
+    GlobalKey<NavigatorState>? navigatorKey,
+    this.initialLocation,
+    this.restorationScopeId,
+    this.observers,
+    this.preload = false,
+  }) : navigatorKey = navigatorKey ?? GlobalKey<NavigatorState>();
+
+  /// The [GlobalKey] to be used by the [Navigator] built for this branch.
+  ///
+  /// A separate Navigator will be built for each StatefulShellBranch in a
+  /// [StatefulShellRoute] and this key will be used to identify the Navigator.
+  /// The routes associated with this branch will be placed o onto that
+  /// Navigator instead of the root Navigator.
+  final GlobalKey<NavigatorState> navigatorKey;
+
+  /// The list of child routes associated with this route branch.
+  final List<RouteBase> routes;
+
+  /// The initial location for this route branch.
+  ///
+  /// If none is specified, the location of the first descendant [GoRoute] will
+  /// be used (i.e. first element in [routes], or a descendant). The default
+  /// location is used when loading the branch for the first time (for instance
+  /// when switching branch using the goBranch method in
+  /// [StatefulShellBranchState]).
+  final String? initialLocation;
+
+  /// Whether this route branch should be preloaded when the associated
+  /// [StatefulShellRoute] is visited for the first time.
+  ///
+  /// If this is true, this branch will be preloaded by navigating to
+  /// the initial location (see [initialLocation]). The primary purpose of
+  /// branch preloading is to enhance the user experience when switching
+  /// branches, which might for instance involve preparing the UI for animated
+  /// transitions etc. Care must be taken to **keep the preloading to an
+  /// absolute minimum** to avoid any unnecessary resource use.
+  final bool preload;
+
+  /// Restoration ID to save and restore the state of the navigator, including
+  /// its history.
+  final String? restorationScopeId;
+
+  /// The observers for this branch.
+  ///
+  /// The observers parameter is used by the [Navigator] built for this branch.
+  final List<NavigatorObserver>? observers;
+
+  @override
+  bool operator ==(Object other) {
+    if (identical(other, this)) {
+      return true;
+    }
+    if (other is! StatefulShellBranch) {
+      return false;
+    }
+    return other.navigatorKey == navigatorKey &&
```

Comment:
This does not check for routes?

Reply:
this class probably should not override the equal method since this is set once in the GoRouter object.

Reply:
This was added as part of supporting "dynamic branches", which now will be handled via future support for adding, updating and removing routes. But regardless, to support change in the branch configuration of a StatefulShellRoute, there needs to be functionality to determine if two branch objects are identical based on data and not only on object identity. But since this isn't used right a the moment, I guess we can remove it for now and re-add it later on.